### ISSUE TITLE
Subscribe lost connection does not error/complete

### ISSUE THREAD

**Issue by taylormsj** [apollographql/apollo-client#5115]
**Intended outcome:** 
When a websocket disconnects it tries to reconnect and eventually gives up. I would expect error or complete to be called so my app knows the connection is lost and can't be recovered. 

**Actual outcome:** 
Nothing happens (apart from console warnings and errors)

**How to reproduce the issue:**
Create new `ApolloClient` with appropriate `WebSocketLink` that has `options: { reconnect: true }`, then subscribe.

```
client
  .subscribe({
    query
  })
  .subscribe({
    start(){
      console.log('Start');
    },
    next(data){
      console.log("Next: ", data);
    },
    error(e){
      console.log('Error: ', e);
    },
    complete(){
      console.log('Complete');
    }
  });
```

`next` is working as expected.
`start`, `error` & `complete` are never called.

**Comment by ksrb** [comment]
After much googling it seems that websocket lifecycle events can be listened to on the `WebSocketLink`'s `subscriptionClient`.

See: https://github.com/apollographql/subscriptions-transport-ws/issues/381#issuecomment-381381568

For Typescript users:
```tsx
import ApolloClient from "apollo-client";
import { HttpLink } from "apollo-link-http";
import { InMemoryCache } from "apollo-cache-inmemory";
import { WebSocketLink } from "apollo-link-ws";
import { getMainDefinition } from "apollo-utilities";
import { split } from "apollo-link";

import { graphqlUrl, graphqlWebSocket } from "constants/urls";

const httpLink = new HttpLink({ uri: graphqlUrl });

const wsLink = new WebSocketLink({
  uri: graphqlWebSocket,
  options: {
    lazy: true,
    reconnect: true,
  },
});

// @ts-ignore at the time of writing  the subscriptionClient is private in the type definition
wsLink.subscriptionClient.on("connecting", () => {
  console.log("connecting");
});

// @ts-ignore
wsLink.subscriptionClient.on("connected", () => {
  console.log("connected");
});

// @ts-ignore
wsLink.subscriptionClient.on("reconnecting", () => {
  console.log("reconnecting");
});

// @ts-ignore
wsLink.subscriptionClient.on("reconnected", () => {
  console.log("reconnected");
});

// @ts-ignore
wsLink.subscriptionClient.on("disconnected", () => {
  console.log("disconnected");
});

const link = split(
  ({ query }) => {
    const definition = getMainDefinition(query);
    return (
      definition.kind === "OperationDefinition" &&
      definition.operation === "subscription"
    );
  },
  wsLink,
  httpLink,
);

export const client = new ApolloClient({ link, cache: new InMemoryCache() });
```
 
This definitely does not feel like a ideal solution but it's better than nothing 😕 .

**Comment by ucallQA** [comment]
Getting the same problem, except complete() never fires for mine.... I can't do the solution with subscriptionClient because I have a separate setup to do a handshake with AWS

**Comment by alessbell** [comment]
Hi all 👋 Would anyone be able to share a runnable reproduction here?

For convenience, we have a [CodeSandbox](https://codesandbox.io/s/github/apollographql/react-apollo-error-template) and forkable [GitHub repo](https://github.com/apollographql/react-apollo-error-template) with a `subscriptions.tsx` already configured.

We also shipped a recent fix in `3.7.11` which improves WebSocket error handling via https://github.com/apollographql/apollo-client/pull/10586.

Thanks!

**Comment by alessbell** [comment]
Hi @taylormsj - I've confirmed that `onError` and `onCompleted` fire in my own reproductions with recent versions of Apollo Client, so without more information I'll close this out. Please open a new issue for any future bug reports.
